Sources/AblyChat/DefaultConnection.swift

```diff
@@ -0,0 +1,144 @@
+import Ably
+
+// I'm not too keen on this class in the way it is... I had a few difficulties keeping a mutable status and error on the class (due to Sendable conformance). I had to resort to using an actor to manage the status and error. This then meant needing to change the `Connection` protocol so `state` and `error` have async getters. To make things more complicated, Swift supports async getters but not async setters (and it doesn't allow you to mix a sync setter with an async getter). You call the relevant methods on the actor to update the status and error. We should revisit this as part of wider concurrency concerns here: https://github.com/ably-labs/ably-chat-swift/issues/49
+internal final class DefaultConnection: Connection {
+    // (CHA-CS2a) The chat client must expose its current connection status.
+    internal var status: ConnectionStatus {
+        get async { await connectionStatusManager.status }
+    }
+
+    // (CHA-CS2b) The chat client must expose the latest error, if any, associated with its current status.
+    internal var error: ARTErrorInfo? {
+        get async { await connectionStatusManager.error }
+    }
+
+    private let realtime: any RealtimeClientProtocol
+    private let timerManager = TimerManager()
+    private let connectionStatusManager = ConnectionStatusManager()
+
+    internal init(realtime: any RealtimeClientProtocol) {
+        // (CHA-CS3) The initial status and error of the connection will be whatever status the realtime client returns whilst the connection status object is constructed.
+        self.realtime = realtime
+        Task {
+            await connectionStatusManager.updateStatus(to: .init(from: realtime.connection.state))
+            await connectionStatusManager.updateError(to: realtime.connection.errorReason)
+        }
+    }
+
+    // (CHA-CS4d) Clients must be able to register a listener for connection status events and receive such events.
+    internal func onStatusChange(bufferingPolicy: BufferingPolicy) -> Subscription<ConnectionStatusChange> {
+        let subscription = Subscription<ConnectionStatusChange>(bufferingPolicy: bufferingPolicy)
+
+        // (CHA-CS5) The chat client must monitor the underlying realtime connection for connection status changes.
+        realtime.connection.on { [weak self] stateChange in
+            guard let self else {
+                return
+            }
+            let currentState = ConnectionStatus(from: stateChange.current)
+            let previousState = ConnectionStatus(from: stateChange.previous)
+
+            Task {
+                if await status == currentState {
+                    return
```

Comment:
It returns from the task, is it intentional?

Reply:
ahh good catch... redundant check tbh - removed it.

CHA-CS5b states all emitted events should be forwarded so it's probably best I don't omit this just because it's the same. We can just trust we wouldn't get the same